Implement unified API for describing addresses (name+offset format)

IndAlok opened the pull request:
<!-- Filling this template is mandatory -->

**Your checklist for this pull request**
- [x] I've read the [guidelines for contributing](https://github.com/rizinorg/rizin/blob/dev/CONTRIBUTING.md) to this repository.
- [x] I made sure to follow the project's [coding style](https://github.com/rizinorg/rizin/blob/dev/DEVELOPERS.md#code-style).
- [x] I've documented every `RZ_API` function and struct this PR changes.
- [x] I've added tests that prove my changes are effective (required for changes to `RZ_API`).
- [ ] I've updated the [Rizin book](https://github.com/rizinorg/book) with the relevant information (if needed).
- [ ] I've used AI tools to generate fully or partially these code changes and I'm sure the changes are not copyrighted by somebody else.

**Detailed description**

This PR introduces a unified API for describing addresses in a human-readable format (e.g., `function_name+0x10`), addressing the fragmentation issue where multiple parts of the codebase implemented similar functionality in slightly different ways.

### Problem

As discussed in rizinorg/cutter#3417 and rizinorg/cutter#3416, Rizin had several places implementing address description logic:
- `asm.reloff` flag handling in disassembly
- Stacktrace/backtrace printing
- Cross-reference (`axt`) output
- `rz_print_offset_sg` in cmd_print.c
- `print_addr` in print.c

These implementations were tightly coupled to their specific use cases and couldn't be easily reused by other parts of Rizin or external tools like Cutter.

### Solution

A new unified API in `librz/core/caddr.c` provides:

**Core Types:**
- `RzAddrDescribeOptions` - Configuration flags for customizing output behavior
- `RzAddrDescription` - Comprehensive address metadata structure

**Main API Functions:**
- `rz_core_addr_describe()` - Get full address description with all metadata
- `rz_core_addr_get_name_delta()` - Simple "name+offset" string (most common use case)
- `rz_core_addr_description_to_string()` - Format description as customizable string
- `rz_core_addr_describe_pj()` - JSON output for command handlers
- `rz_core_addr_description_free()` - Cleanup function

**Key Features:**
- Uses flags/options structure instead of relying on global config options
- Returns structured data that can be formatted differently in different contexts
- Supports function names, flag names, and source line information
- Provides both hex and decimal offset formatting
- JSON output support for commands
- Reusable by Cutter and other external tools

### Files Changed

**New Files:**
- `librz/core/caddr.c` - Core implementation (~620 lines)
- `test/unit/test_core_addr.c` - Unit tests (6 test cases)
- `doc/address_description_api.md` - API documentation

**Refactored to use new API:**
- `librz/core/core.c` - `getNameDelta()` callback
- `librz/core/cdebug.c` - Backtrace info generation
- `librz/core/cprint.c` - Hexdump address display
- `librz/core/cmd/cmd_analysis.c` - Cross-references listing
- `librz/core/cmd/cmd_debug.c` - Backtrace and trace printing
- `librz/core/cmd/cmd_seek.c` - Seek history display
- `librz/core/tui/visual.c` - Visual mode xrefs
- `librz/core/tui/vmenus_graph.c` - Graph menu xrefs

**Test plan**

### Unit Tests

Run the new unit tests:

```bash
meson test -C build --suite unit
```
Specific test for the new API:

```bash
# Build with tests enabled
meson setup build -Denable_tests=true
ninja -C build

# Run the core_addr test
meson test -C build -v core_addr
```
Expected output:

```bash
1/6 test_rz_core_addr_describe_options_new      OK
2/6 test_rz_core_addr_describe_basic            OK
3/6 test_rz_core_addr_description_to_string     OK
4/6 test_rz_core_addr_with_function             OK
5/6 test_rz_core_addr_get_name_delta            OK
6/6 test_rz_core_addr_describe_pj               OK
```

### Manual Testing

1. **Test basic address description:**
```bash
rizin -A /bin/ls
[0x00004000]> afl~main
[0x00004000]> s main+0x10
[0x00004010]> axt
```
The output should now show function context using the unified API.

2. **Test seek history:**
```bash
rizin /bin/ls
[0x00000000]> aa
[0x00000000]> s main
[0x00004000]> s main+0x20
[0x00004020]> sh
```
Seek history should show addresses with name+offset format.

3. **Test backtrace (requires debugging):**
```bash
rizin -d /bin/ls
[0x7f...]> dc
[0x7f...]> dbt
```
Backtrace should use the new API for address formatting.

...

**Closing issues**

Closes #4993

...

notxvilka reviewed (changes requested):
Good job, but:

- Remove all irrelevant changes (e.g. in CI)
- Move documentation to the Doxygen format in the corresponding files

notxvilka reviewed (changes requested):
Please rebase on top of the latest `dev` to fix some CI errors.

notxvilka commented on .github/workflows/ci.yml:
```diff
@@ -118,7 +118,7 @@ jobs:
             continue-on-error: false
             run_tests: true
-            enabled: ${{ (needs.changes.outputs.edited == 'true' && github.event_name == 'pull_request') }}
+            enabled: ${{ (needs.changes.outputs.edited == 'true' || github.event_name == 'pull_request') }}
```
Wrong changes again

IndAlok replied:
Pardon for that, has been reverted. I misinterpreted the CI failure cause to be the && compulsion .

wargio commented on librz/core/caddr.c:
```diff
@@ -0,0 +1,615 @@
+// SPDX-FileCopyrightText: 2024 RizinOrg <info@rizin.re>
```
```suggestion
// SPDX-FileCopyrightText: 2025 RizinOrg <info@rizin.re>
```

wargio commented on librz/core/caddr.c:
```diff
@@ -0,0 +1,615 @@
+ * and compute the delta.
+ */
+static void addr_describe_from_flags(RzCore *core, RzAddrDescription *desc, ut64 addr, const RzAddrDescribeOptions *opts) {
+	rz_return_if_fail(core && desc && opts);
```
if you have an assert, you should add RZ_NONNULL/NULLABLE to the pointers.

wargio commented on librz/core/caddr.c:
```diff
@@ -0,0 +1,615 @@
+
+	const char *name = (opts->use_realnames && fi->realname) ? fi->realname : fi->name;
+
+	desc->flag_name = rz_str_dup(name);
```
check for null ptr?

wargio commented on librz/core/caddr.c:
```diff
@@ -0,0 +1,615 @@
+		return;
+	}
+
+	desc->fcn_name = rz_str_dup(fcn->name);
```
check for null ptr

wargio commented on librz/core/caddr.c:
```diff
@@ -0,0 +1,615 @@
+ * Helper function to find a function containing the address
+ * and compute the delta from the function start.
+ */
+static void addr_describe_from_function(RzCore *core, RzAddrDescription *desc, ut64 addr, const RzAddrDescribeOptions *opts) {
```
this function should return true or false.

wargio replied:
same as above, add NULLABLE/NONNULL

wargio commented on librz/core/caddr.c:
```diff
@@ -0,0 +1,615 @@
+ * Helper function to find a flag at or near the given address
+ * and compute the delta.
+ */
+static void addr_describe_from_flags(RzCore *core, RzAddrDescription *desc, ut64 addr, const RzAddrDescribeOptions *opts) {
```
should return true or false

wargio commented on librz/core/caddr.c:
```diff
@@ -0,0 +1,615 @@
+ * Helper function to retrieve source file and line number
+ * from debug information if available.
+ */
+static void addr_describe_from_source(RzCore *core, RzAddrDescription *desc, ut64 addr, const RzAddrDescribeOptions *opts) {
```
should return true or false & same as above, add NULLABLE/NONNULL

wargio commented on librz/core/caddr.c:
```diff
@@ -0,0 +1,615 @@
+	}
+
+	if (sample->file) {
+		desc->source_file = rz_str_dup(sample->file);
```
check for null ptr

wargio commented on librz/core/caddr.c:
```diff
@@ -0,0 +1,615 @@
+	addr_describe_from_flags(core, desc, addr, opts);
+
+	// Get source line information if requested
+	addr_describe_from_source(core, desc, addr, opts);
```
check for failure

wargio commented on librz/core/caddr.c:
```diff
@@ -0,0 +1,615 @@
+		name = desc->fcn_name;
+		delta = desc->fcn_delta;
+		has_name = true;
+	} else if (opts->show_flag && desc->flag_name) {
```
use RZ_STR_EMPTY

wargio commented on librz/core/caddr.c:
```diff
@@ -0,0 +1,615 @@
+RZ_API void rz_core_addr_describe_pj(RZ_NONNULL RzCore *core, RZ_NONNULL PJ *pj, ut64 addr, RZ_NULLABLE const RzAddrDescribeOptions *opts) {
+	rz_return_if_fail(core && pj);
+
+	RzAddrDescription *desc = rz_core_addr_describe(core, addr, opts);
```
this should be an api to translate a given `RzAddrDescription` into a json.

IndAlok committed:
core: Add unified API for describing addresses (resolves rizinorg#4993)

IndAlok committed:
core: Add unified API for describing addresses (resolves rizinorg#4993)

notxvilka commented on .github/workflows/ci.yml:
```diff
@@ -120,8 +120,6 @@ jobs:
             run_tests: true
             enabled: ${{ (needs.changes.outputs.edited == 'true' && github.event_name == 'pull_request') }}
             timeout: 120
-            # The existing leaks in Rizin will make the tests fail otherwise
```
Still unnecessary changes again. Please do not change irrelevant files!

IndAlok replied:
My bad, assure you that was an accidental documentation purge I didn't intend, won't happen again. Reverted.

notxvilka replied:
You still have changes in this file even after the latest commit.

IndAlok replied:
Reverted the misformatting as well. Thanks for informing, as aforementioned, won't happen in future.

IndAlok committed:
Reverted mistakenly purged comment on the workflow

IndAlok committed:
Reverted mistakenly purged comment on the workflow

IndAlok committed:
Reverted unintended misformatting in CI workflow

IndAlok committed:
Reverted unintended misformatting in CI workflow

IndAlok committed:
core: Fix pxW and prompt tests by adding backward-compatible delta limit

This commit fixes the test failures caused by behavioral changes in the
unified address description API:

1. pxW tests failing (cmd_pxw, helloworld-gcc-elf):
   - The old pxW code had a hardcoded 8192 limit for flag delta
   - Added max_flag_delta option to RzAddrDescribeOptions
   - rz_core_addr_get_flag_offset() now uses max_flag_delta=-1 (8192 limit)

2. prompt settings test failing (test/db/tools/rz):
   - The old prompt code used spaces around +/- (e.g., 'entry0 + 1')
   - Added use_spaces_around_delta option to RzAddrDescribeOptions
   - Added rz_core_addr_get_flag_offset_prompt() for prompt display

Files changed:
- librz/include/rz_core.h: Added max_flag_delta and use_spaces_around_delta
- librz/core/caddr.c: Implemented delta limit and spacing options
- librz/core/core.c: Use new prompt function
- test/unit/test_core_addr.c: Added tests for new options

IndAlok committed:
core: Fix pxW and prompt tests by adding backward-compatible delta limit

This commit fixes the test failures caused by behavioral changes in the
unified address description API:

1. pxW tests failing (cmd_pxw, helloworld-gcc-elf):
   - The old pxW code had a hardcoded 8192 limit for flag delta
   - Added max_flag_delta option to RzAddrDescribeOptions
   - rz_core_addr_get_flag_offset() now uses max_flag_delta=-1 (8192 limit)

2. prompt settings test failing (test/db/tools/rz):
   - The old prompt code used spaces around +/- (e.g., 'entry0 + 1')
   - Added use_spaces_around_delta option to RzAddrDescribeOptions
   - Added rz_core_addr_get_flag_offset_prompt() for prompt display

Files changed:
- librz/include/rz_core.h: Added max_flag_delta and use_spaces_around_delta
- librz/core/caddr.c: Implemented delta limit and spacing options
- librz/core/core.c: Use new prompt function
- test/unit/test_core_addr.c: Added tests for new options

notxvilka reviewed (commented):
The crux of the change looks good. I am not sure about naming, if `RzAddr*` and `rz_core_addr_*` names are best. Lets see what others think on this matter.

IndAlok committed:
Update librz/core/caddr.c

Co-authored-by: Giovanni <561184+wargio@users.noreply.github.com>

IndAlok committed:
Update librz/core/caddr.c

Co-authored-by: Giovanni <561184+wargio@users.noreply.github.com>

IndAlok committed:
core/caddr: Add proper error handling and API improvements

- Add RZ_NONNULL annotations to helper function parameters
- Add null pointer checks after rz_str_dup() to detect allocation failures
- Change helper functions to return bool for error propagation
- Handle memory allocation failures in rz_core_addr_describe() by
  checking helper return values and cleaning up on failure
- Use RZ_STR_ISEMPTY macro for consistent empty string checks
- Extract rz_core_addr_description_to_pj() as separate API for
  converting RzAddrDescription to JSON format

IndAlok committed:
core/caddr: Add proper error handling and API improvements

- Add RZ_NONNULL annotations to helper function parameters
- Add null pointer checks after rz_str_dup() to detect allocation failures
- Change helper functions to return bool for error propagation
- Handle memory allocation failures in rz_core_addr_describe() by
  checking helper return values and cleaning up on failure
- Use RZ_STR_ISEMPTY macro for consistent empty string checks
- Extract rz_core_addr_description_to_pj() as separate API for
  converting RzAddrDescription to JSON format

IndAlok committed:
Merge branch 'dev' of https://github.com/IndAlok/rizin into dev

wargio commented on librz/include/rz_core.h:
```diff
@@ -595,6 +638,61 @@ RZ_API RzCmdStatus rz_core_debug_plugins_print(RZ_NONNULL RZ_BORROW RzCore *core
 RZ_API void rz_core_debug_map_print(RzCore *core, ut64 addr, RzCmdStateOutput *state);
 
+/**
+ * \defgroup caddr Address Description API
+ * \brief Unified API for describing addresses in human-readable formats
+ *
+ * This API provides functions to generate human-readable descriptions of memory
+ * addresses, combining information from flags, functions, and debug symbols.
+ * It supports various output formats including name+offset notation (e.g., "main+0x10"),
+ * source file information, and JSON output.
+ *
+ * Example usage:
+ * \code
+ * char *desc = rz_core_addr_get_name_delta(core, 0x401010);
+ * // Returns "main+16" or "0x401010" if no symbol found
+ * rz_cons_printf("Address: %s\n", desc);
+ * free(desc);
+ * \endcode
+ *
+ * @{
+ */
```
i would create a specific header for this and then add the core api here for its usage, otherwise the structure should be called `RzCoreAddrDescribe`

wargio replied:
also the name of this structure is still undecided for us (rizin core team)

IndAlok replied:
Till the finalization, I think as suggested, the naming could be according to 

> RzCoreAddrDescribe

wargio commented on librz/include/rz_core.h:
```diff
@@ -595,6 +638,61 @@ RZ_API RzCmdStatus rz_core_debug_plugins_print(RZ_NONNULL RZ_BORROW RzCore *core
+/** \brief Get flag-relative offset string with spaces for prompts (e.g., "sym.func + 0x10") */
+RZ_API RZ_OWN char *rz_core_addr_get_flag_offset_prompt(RZ_NONNULL RzCore *core, ut64 addr);
+
+/** @} */ // end of caddr group
```
not needed

```suggestion
```

wargio commented on librz/include/rz_core.h:
```diff
@@ -595,6 +638,61 @@ RZ_API RzCmdStatus rz_core_debug_plugins_print(RZ_NONNULL RZ_BORROW RzCore *core
+ * free(desc);
+ * \endcode
+ *
+ * @{
```
not needed
```suggestion
```

wargio commented on librz/include/rz_core.h:
```diff
@@ -595,6 +638,61 @@ RZ_API RzCmdStatus rz_core_debug_plugins_print(RZ_NONNULL RZ_BORROW RzCore *core
+/** \brief Get flag-relative offset string (e.g., "sym.func+0x10") */
+RZ_API RZ_OWN char *rz_core_addr_get_flag_offset(RZ_NONNULL RzCore *core, ut64 addr);
+/** \brief Get flag-relative offset string with spaces for prompts (e.g., "sym.func + 0x10") */
+RZ_API RZ_OWN char *rz_core_addr_get_flag_offset_prompt(RZ_NONNULL RzCore *core, ut64 addr);
```
the comments should go in the `.c` file, not the headers.

IndAlok committed:
as per suggestion

Co-authored-by: Giovanni <561184+wargio@users.noreply.github.com>

IndAlok committed:
as per suggestion

Co-authored-by: Giovanni <561184+wargio@users.noreply.github.com>

IndAlok committed:
core/caddr: Rename structs to RzCoreAddr* and clean up header comments

- Rename RzAddrDescribeOptions to RzCoreAddrDescribeOptions

- Rename RzAddrDescription to RzCoreAddrDescription

- Remove inline brief comments from header file declarations

- Keep full doxygen documentation in .c file

IndAlok committed:
Merge branch 'dev' of https://github.com/IndAlok/rizin into dev

IndAlok committed:
fix: clang-formating for core.c and rz_core.h

IndAlok committed:
Merge branch 'dev' into dev

IndAlok committed:
oops

IndAlok committed:
Version 18 (cry)

notxvilka commented on subprojects/packagefiles/libzip-1.11.3/meson.build:
```diff
@@ -82,6 +82,14 @@ foreach fcn : underscore_functions
+# Enable fdopen support if fdopen or _fdopen is available
+# This is needed for zip_fdopen.c to compile without unused variable warnings
+# Check for fdopen (POSIX/Linux/macOS) or _fdopen (Windows)
+if cc.has_function('fdopen') or cc.has_function('fdopen', prefix: '#include <stdio.h>') or cc.has_function('_fdopen') or cc.has_function('_fdopen', prefix: '#include <stdio.h>')
```
Extract this change into a separate PR, please name the branch `dist-*`

Rot127 commented on librz/core/cmd/cmd_debug.c:
```diff
@@ -1066,34 +1066,15 @@ static void backtrace_vars(RzCore *core, RzList /*<RzDebugFrame *>*/ *frames) {
-					"%s", fi->name);
-			}
-		}
+		char *flagdesc = rz_core_addr_get_flag_offset(core, f->addr);
```
Please don't pass `core` here. Just pass the `RzFlag` instance. We want to reduce the dependencies on COre, not increase them.

Rot127 replied:
For this I would also suggest move the function into the `rz_flag` module.

Rot127 commented on librz/core/caddr.c:
```diff
@@ -0,0 +1,737 @@
+ * \param opts Options controlling behavior
+ * \return true on success (including when no function found), false on error (e.g., memory allocation failure)
+ */
+static bool addr_describe_from_function(RZ_NONNULL RzCore *core, RZ_NONNULL RzCoreAddrDescription *desc, ut64 addr, RZ_NONNULL const RzCoreAddrDescribeOptions *opts) {
```
I these case I would just pass the `RzAnalysis` instance instead of `core` again.

Rot127 commented on librz/core/caddr.c:
```diff
@@ -0,0 +1,737 @@
+		return true; // Not requested, not an error
+	}
+
+	RzBinFile *bf = rz_bin_cur(core->bin);
```
Pass the `RzBinFile` instance directly. Not core

Rot127 commented on librz/core/caddr.c:
```diff
@@ -0,0 +1,737 @@
+ * \param opts Display options
+ * \return Newly allocated formatted string. Caller must free.
+ */
+RZ_API RZ_OWN char *rz_core_addr_format_for_display(RZ_NONNULL RzCore *core, ut64 addr, RZ_NULLABLE const RzCoreAddrDescribeOptions *opts) {
```
`RzPrint` instead of core

Rot127 commented on librz/include/rz_core.h:
```diff
@@ -200,6 +200,49 @@ typedef struct {
 } RzCoreGadget;
 
+/**
+ * \brief Options for controlling address description output
```
```suggestion
 * \brief Options for controlling address description output.
```

Please be more pedantic with spelling and grammar in API docs.

Rot127 commented on librz/include/rz_core.h:
```diff
@@ -595,6 +638,42 @@ RZ_API RzCmdStatus rz_core_debug_plugins_print(RZ_NONNULL RZ_BORROW RzCore *core
+ *
+ */
+
+RZ_API RZ_OWN RzCoreAddrDescribeOptions *rz_core_addr_describe_options_new(void);
```
These declarations still have to stay in `rz_core` because `RzFlag` cannot depend on `RzAnalysis` etc, but if the functions don't use `RzCore`, it is easier to move it out at some point in the future, into a new `RzUI` module or similar.

Rot127 reviewed (changes requested):
Added some comments regarding the naming and design.
Please wait though what @notxvilka and @wargio say to it.

IndAlok committed:
Spelling

Co-authored-by: Rot127 <45763064+Rot127@users.noreply.github.com>

IndAlok committed:
Merge branch 'dev' into dev

IndAlok committed:
core/caddr: reduce RzCore dependencies in helper functions

Refactor internal helper functions and some public APIs to take specific
subsystem instances instead of RzCore, reducing coupling:

- addr_describe_from_flags(): take RzFlag* instead of RzCore*
- addr_describe_from_function(): take RzAnalysis* instead of RzCore*
- addr_describe_from_source(): take RzBinFile* instead of RzCore*
- rz_core_addr_format_for_display(): take RzPrint* instead of RzCore*
- rz_core_addr_get_flag_offset(): take RzFlag* instead of RzCore*
- rz_core_addr_get_flag_offset_prompt(): take RzFlag* instead of RzCore*

This makes the code more modular and easier to refactor in the future,
as suggested in code review.

notxvilka reviewed (commented):
@wargio suggested using `RzOffset` naming, but after discussing, I think `RzCoreAddr` is fine as well. No strong opinion on that. For me it's ok as is.

wargio commented:
> @wargio suggested using `RzOffset` naming, but after discussing, I think `RzCoreAddr` is fine as well. No strong opinion on that. For me it's ok as is.

Ok for me too. @IndAlok  rename the structure to `RzCoreAddr`

IndAlok committed:
core/caddr: rename structures to RzCoreAddr and RzCoreAddrOptions

As requested by maintainers, rename the structures and related functions:

Structures:
- RzCoreAddrDescription -> RzCoreAddr
- RzCoreAddrDescribeOptions -> RzCoreAddrOptions

Functions:
- rz_core_addr_describe_options_new() -> rz_core_addr_options_new()
- rz_core_addr_describe_options_free() -> rz_core_addr_options_free()
- rz_core_addr_description_free() -> rz_core_addr_free()
- rz_core_addr_description_to_string() -> rz_core_addr_to_string()
- rz_core_addr_description_to_pj() -> rz_core_addr_to_pj()

IndAlok committed:
Merge branch 'dev' into dev

IndAlok committed:
Merge branch 'dev' into dev
